tls/tests/conftest.py

```diff
@@ -1,16 +1,80 @@
 # (C) Datadog, Inc. 2019-present
 # All rights reserved
 # Licensed under a 3-clause BSD style license (see LICENSE)
+import os
+
 import pytest
+from six import iteritems
 
+from datadog_checks.dev import TempDir, docker_run
 from datadog_checks.tls.utils import days_to_seconds
 
 from .utils import download_cert, temp_binary
 
+try:
+    from contextlib import ExitStack
+except ImportError:
+    from contextlib2 import ExitStack
+
+
+HERE = os.path.dirname(os.path.abspath(__file__))
+
+CA_CERT = os.path.join(HERE, 'compose', 'ca.crt')
+
 
 @pytest.fixture(scope='session', autouse=True)
 def dd_environment():
-    yield {'server': 'https://www.google.com'}
+    with docker_run(
+        os.path.join(HERE, 'compose', 'docker-compose.yml'), build=True,
+    ):
+        e2e_metadata = {'docker_volumes': ['{}:{}'.format(CA_CERT, CA_CERT)]}
+        yield {'server': 'valid.mock'}, e2e_metadata
+
+
+@pytest.fixture(scope='session', autouse=True)
+def mock_dns():
+    import socket
+
+    _orig_getaddrinfo = socket.getaddrinfo
+
+    def patched_getaddrinfo(host, *args, **kwargs):
+        if host.endswith('.mock'):
+            # nginx doesn't support multiple tls versions from the same container
+            if 'tls-v1-1' in host:
+                port = 4444
+            elif 'tls-v1-2' in host:
+                port = 4445
+            elif 'tls-v1-3' in host:
+                port = 4446
+            else:
+                port = 4443
+
+            # See socket.getaddrinfo, just updating the hostname here.
+            return [(2, 1, 6, '', ('127.0.0.1', port))]
+
+        return _orig_getaddrinfo(host, *args, **kwargs)
+
+    socket.getaddrinfo = patched_getaddrinfo
+    yield
+    socket.getaddrinfo = _orig_getaddrinfo
+
+
+@pytest.fixture(scope='session', autouse=True)
+def certs(dd_environment, mock_dns):
+    downloads = {'https://valid.mock': 'valid.pem', 'https://expired.mock': 'expired.pem'}
+    raw_downloads = {
+        'https://valid.mock': 'valid.crt',
+    }
+    certs = {}
+    with ExitStack() as stack:
+        tmp_dir = stack.enter_context(TempDir())
+        for address, name in iteritems(downloads):
+            filepath = os.path.join(tmp_dir, name)
+            certs[name] = stack.enter_context(download_cert(filepath, address))
+        for address, name in iteritems(raw_downloads):
+            filepath = os.path.join(tmp_dir, name)
+            certs[name] = stack.enter_context(download_cert(filepath, address, raw=True))
+        yield certs
```

Comment:
1. Seems `download_cert` doesn't need to be a context manager anymore since the TempDir is created outside.

2. `download_cert` probably don't need to return the `filepath` since we are passing it.

3. If we do 1., the `with ExitStack` can be replaced with `with TempDir()`

Reply:
Good idea, `download_cert` was too close to the original design. This is much better 👍